### PR TITLE
Autograph guide

### PR DESCRIPTION
**Pull request by josh146**

### CODE REVIEW

File: doc/dev/autograph.rst [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+        return x ** 3
+
+will not be correctly captured by AutoGraph, and instead will be interpreted as
+
+.. code-block:: python
+
+    def f(x):
+        if x > 5:
+            x = x ** 2
+        return x ** 3
```
Comment (josh146) [inline doc/dev/autograph.rst:195]: @dime10 it seems that TF doesn't have this issue, so this might actually be a bug 🤔

Comment (josh146) [reply doc/dev/autograph.rst:195]: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/autograph/g3doc/reference/control_flow.md#return-statements

Comment (josh146) [reply doc/dev/autograph.rst:195]: I might rewrite this section and simply say 'Return statements inside `if` statements are not yet supported. No error will occur, but the resulting function will not have the expected behaviour'.

Comment (dime10) [reply doc/dev/autograph.rst:195]: Yes that's right, it just hasn't been implemented yet. There are few other features that we don't have as well, like `break`

File: doc/dev/autograph.rst [hunk 2]
```diff
@@ -0,0 +1,643 @@
...
+compile time):
+
+>>> @qjit(autograph=True)
+... def f(n):
+...     x = - jnp.log(n)
+...     for k in range(1, n + 1):
+...         x = x + 1 / k
+...     return x
+>>> f(100000)
+array(0.57722066)
```
Comment (josh146) [inline doc/dev/autograph.rst:354]: Guess the constant 😆

**Review by dime10 — CHANGES_REQUESTED** [review]
Phenomenal guide @josh146!! 🚀 🚀 🚀 

Easy to follow and still provides the necessary information 💪

File: doc/dev/autograph.rst [hunk 3]
```diff
@@ -0,0 +1,643 @@
...
+        params = ["0", "1", "2"]
+        for x in params:
+            qml.RY(int(x) * jnp.pi / 4, wires=0)
+        return qml.expval(qml.PauliZ(0))
+
+>>> f()
+array(-0.70710678)
+
+The Python ``range`` function is also fully supported by AutoGraph, even when
+its input is a **dynamic variable** (e.g., its numeric value is only known at
```
Comment (dime10) [inline doc/dev/autograph.rst:344]: i.e. ? :)
```suggestion
its input is a **dynamic variable** (i.e., its numeric value is only known at
```

File: doc/dev/autograph.rst [hunk 4]
```diff
@@ -0,0 +1,643 @@
...
+
+                @cond(x[j] > 0)
+                def trainable_gate():
+                    qml.RX(x[j], wires=j)
+
+                @trainable_gate.else_if(x[j] < 0)
+                def negative_gate():
+                    qml.RY(x[j], wires=j)
+
+                trainable_gate.otherwise(lambda: None)
```
[the remark this answers is not in the document]
Comment (josh146) [reply doc/dev/autograph.rst:95]: No I had to include this, I can't remember why (I was getting an error message?) This was on 0.3 though, not sure if something changed

File: doc/dev/autograph.rst [hunk 5]
```diff
@@ -0,0 +1,643 @@
...
+
+``while`` loops are currently not supported.
+
+Nested functions
+----------------
+
+AutoGraph will continue to work even when the qjit-compiled function
+itself calls nested functions. All functions called within the
+qjit-compiled function will also have Python control flow captured
+and converted by AutoGraph.
```
[the remark this answers is not in the document]
Comment (josh146) [reply doc/dev/autograph.rst:124]:
```suggestion
AutoGraph will continue to work even when the qjit-compiled function
itself calls nested functions. All functions called within the
qjit-compiled function will also have Python control flow captured
and converted by AutoGraph.

In addition, built-in functions from `jax`, `pennylane`, and `catalyst`
are automatically *excluded* from the AutoGraph conversion when called
within the qjit-compiled function.
```

File: doc/dev/autograph.rst [hunk 6]
```diff
@@ -0,0 +1,643 @@
...
+
+>>> @qjit(autograph=True)
+... def f(x):
+...     if x > 5:
+...         y = 5.0
+...     else:
+...         y = 4
+...     return y
+>>> f(0.5)
+TypeError: Conditional requires consistent return types across all branches
```
[the remark this answers is not in the document]
Comment (josh146) [reply doc/dev/autograph.rst:227]: My general approach is that the docs should show common pitfalls for users of the software _now_, and we ensure that we update the docs as the software abilities change

File: doc/dev/autograph.rst [hunk 7]
```diff
@@ -0,0 +1,643 @@
...
+
+>>> @qjit(autograph=True)
+... def f(x):
+...     if x > 5:
+...         y = "a"
+...     else:
+...         y = "b"
+...     return y
+>>> f(0.5)
+TypeError: Value 'a' with type <class 'str'> is not a valid JAX type
```
[the remark this answers is not in the document]
Comment (josh146) [reply doc/dev/autograph.rst:298]: Have edited!

File: doc/dev/autograph.rst [hunk 8]
```diff
@@ -0,0 +1,643 @@
...
+        for x in params:
+            qml.RY(int(x) * jnp.pi / 4, wires=0)
+        return qml.expval(qml.PauliZ(0))
+
+>>> f()
+array(-0.70710678)
+
+The Python ``range`` function is also fully supported by AutoGraph, even when
+its input is a **dynamic variable** (e.g., its numeric value is only known at
+compile time):
```
[the remark this answers is not in the document]
Comment (josh146) [reply doc/dev/autograph.rst:345]: oops 🤦

File: doc/dev/autograph.rst [hunk 9]
```diff
@@ -0,0 +1,643 @@
...
+...     x = jnp.array([0.1, 0.2, 0.3])
+...     for i in range(3):
+...         qml.RX(x[i], wires=i)
+...     return qml.expval(qml.PauliZ(0))
+>>> f()
+array(0.99500417)
+
+
+Note that if the object you are indexing **cannot** be converted to a JAX
+array? In this case, it is not possible for AutoGraph to capture this for
```
[the remark this answers is not in the document]
Comment (josh146) [reply doc/dev/autograph.rst:415]: I like the first one!

File: doc/dev/autograph.rst [hunk 10]
```diff
@@ -0,0 +1,643 @@
...
+...     x = [0.0, 1 / 4 * jnp.pi, 2 / 4 * jnp.pi]
+...     for i in range(n):
+...         qml.RY(x[i], wires=0)
+...     return qml.expval(qml.PauliZ(0))
+Warning: If you intended for the conversion to happen, make sure that the(now dynamic) loop variable is not used in tracing-incompatible ways,
+for instance by indexing a Python list with it. In that case, the list should be wrapped into an array.
+To understand different types of JAX tracing errors, please refer to the guide at: https://jax.readthedocs.io/en/latest/errors.html
+If you did not intend for the conversion to happen, you may safely ignore this warning.
+TracerIntegerConversionError: The __index__() method was called on traced array with shape int64[].
+See https://jax.readthedocs.io/en/latest/errors.html#jax.errors.TracerIntegerConversionError
```
[the remark this answers is not in the document]
Comment (josh146) [reply doc/dev/autograph.rst:476]: Yep, I had the same dilemma 🤔 I almost did delete everything except the error, but then I thought -- what if the user searches for the warning text?

File: frontend/catalyst/__init__.py [hunk 11]
```diff
@@ -77,7 +77,81 @@
 from catalyst.utils.exceptions import CompileError
 
 autograph_ignore_fallbacks = False
+"""bool: Specify whether AutoGraph should avoid raising
```
[the remark this answers is not in the document]
Comment (josh146) [reply frontend/catalyst/__init__.py:80]: Yep! Sphinx recognizes it even though I don't think it does anything logically in Python.

File: frontend/catalyst/__init__.py [hunk 12]
```diff
@@ -77,7 +77,81 @@
...
+
 autograph_strict_conversion = False
+"""bool: Specify whether AutoGraph should raise exceptions
+when conversion fails, rather than falling back to interpreting
+control flow by Python at compile-time.
+
+**Example**
+
+In certain cases, AutoGraph will fail to convert control flow(for example,
+when an object that can not be converted to a JAX array is indexed in a
```
[the remark this answers is not in the document]
Comment (josh146) [reply frontend/catalyst/__init__.py:124]:
```suggestion
when an object that cannot be converted to a JAX array is indexed in a
```

Weirdly enough, I think if you emphasize the not (can *not*) it looks okay to write it as two words?

File: doc/dev/autograph.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (josh146) [reply doc/dev/autograph.rst:231]: that is, treat docs like tests :) the docs only update when the functionality updates

File: doc/dev/autograph.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (dime10) [reply doc/dev/autograph.rst:95]: Got it, I don't think anything changed, must be an oversight (or bug) on our part then

File: doc/dev/autograph.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (dime10) [reply doc/dev/autograph.rst:231]: Sounds good, I will make a note on the ticket :)

File: doc/dev/autograph.rst (hunk 10 above)
[the remark this answers is not in the document]
Comment (dime10) [reply doc/dev/autograph.rst:481]: We did show the warning several times at this point in the docs though 😅

File: doc/dev/autograph.rst [hunk 13]
```diff
@@ -0,0 +1,643 @@
...
+
+>>> @qjit(autograph=True)
+... def f(x):
+...     for y in [0, 4, 5]:
+...         x = x + y
+...     return x
+>>> f(4)
+array(13)
+
+You can also utilize temporary variables within a for loop:
```
[the remark this answers is not in the document]
Comment (josh146) [reply doc/dev/autograph.rst:519]: Have addressed this in 395aa304 :)

[the remark this answers is not in the document]
Comment (josh146) [reply doc/dev/autograph.rst:519]: Let me know if you think additional examples are needed here.

File: doc/dev/autograph.rst (hunk 10 above)
[the remark this answers is not in the document]
Comment (josh146) [reply doc/dev/autograph.rst:480]:
```suggestion
TracerIntegerConversionError: The __index__() method was called on traced array with shape int64[].
See https://jax.readthedocs.io/en/latest/errors.html#jax.errors.TracerIntegerConversionError
```

Good enough reason for me!

File: doc/dev/autograph.rst (hunk 13 above)
[the remark this answers is not in the document]
Comment (dime10) [reply doc/dev/autograph.rst:515]: Looks great :)

File: doc/dev/autograph.rst (hunk 2 above)
Comment (dime10) [reply doc/dev/autograph.rst:358]: image

**Review by dime10 — APPROVED** [review]
👨‍🍳 👌

**Pull request merged**